### ISSUE TITLE
Missing hashCode definitions

### ISSUE THREAD

**Issue by fpoli** [viperproject/silicon#723]
The [Scalastyle](http://www.scalastyle.org/) linter complains about missing `hashCode` definitions in Silicon. I'm not sure if we did that intentionally but it seems worth double-checking, adding a comment to Silicon if there is nothing to fix.

> Defining either `equals()` or `hashCode()` in a class without defining the other is a known source of bugs. Usually, when you define one, you should also define the other.
>
> Further Reading: [Scalastyle - EqualsHashCode](https://scalastyle.beautiful-scala.com/rules-1.5.0.html#org_scalastyle_scalariform_EqualsHashCodeChecker)

https://github.com/viperproject/silicon/blob/fd51ab6fba10db51858924575a3f51e385c6d577/src/main/scala/state/Terms.scala#L478-L508

https://github.com/viperproject/silicon/blob/fd51ab6fba10db51858924575a3f51e385c6d577/src/main/scala/state/Terms.scala#L433-L445

https://github.com/viperproject/silicon/blob/fd51ab6fba10db51858924575a3f51e385c6d577/annotation/Terms.example.scala#L849-L878

https://github.com/viperproject/silicon/blob/fd51ab6fba10db51858924575a3f51e385c6d577/src/main/scala/state/Chunks.scala#L150-L157

**Comment by marcoeilers** [comment]
I'm confused. Don't the first two have both explicit equals and hashCode definitions? The other two look potentially problematic, I agree.

**Comment by fpoli** [comment]
> I'm confused. Don't the first two have both explicit equals and hashCode definitions?

Yes, sorry. I just copy-pasted the warnings, but I guess that the linter didn't see the `lazy` methods.
